### PR TITLE
Feat(planner): add optional Bézier trajectory head as swappable planner

### LINKED ISSUE #45: Proposal: Diffusion decoder with flow matching as optional trajectory planner
## Summary

Unlike discrete textual tokens or autoregressive sequence decoding (which can result in jerky, computationally inefficient trajectory predictions), flow matching operates over continuous values directly producing a smooth continuous motion output - as implemented in the Alpamayo model. This proposal seeks to add a swappable module in the trajectory planner to use flow matching methodology.

## Key Benefits

- Continuous Trajectories: It defines a probability path (a velocity field) connecting noise to the final trajectory plan.
- Geometry-Aware Control: By bypassing discrete motion tokens, the decoder preserves scalar geometry and kinematic constraints natively.
- Real-time Latency: The optimization streamlines both training and inference, achieving single-digit millisecond inference times necessary for safety
- Continuous Motion Output: The flow-matching head regresses the velocity vectors to map the straightest possible path from the starting state to the optimal, multi-modal future trajectory.

## Implementation Proposal
To integrate a flow-matching based trajectory decoder into a Bird's-Eye-View (BEV) feature module, we couldt transform the 2D spatially grid-aligned BEV features into an information-dense conditional context vector. This vector would then guide a continuous Vector Field (Velocity Field) Regressor to denoise raw gaussian noise into smooth, kinematically sound coordinates.

### Architectural Changes
```
BEV Feature Map] ────> [Conv2D / Flatten] ────┐
(B, C, H, W)                                  │
                                              ▼
[Noisy Actions u_t] ──> [Action Projection] ──> [Cross-Attention Layer] ──> [AdaLN (Time t)] ──> [Velocity Field Output v_θ]
(B, T, 2)                                                                                       (B, T, 2)

```
```python
class TimestepEmbedding(nn.Module):
    """Maps scalar timesteps t to high-dimensional sinusoidal embeddings."""
    def __init__(self, dim):
        super().__init__()
        self.dim = dim

    def forward(self, t):
        device = t.device
        half_dim = self.dim // 2
        embeddings = math.log(10000) / (half_dim - 1)
        embeddings = torch.exp(torch.arange(half_dim, device=device) * -embeddings)
        embeddings = t[:, None] * embeddings[None, :]
        embeddings = torch.cat((embeddings.sin(), embeddings.cos()), dim=-1)
        return embeddings

class FlowMatchingControlDecoder(nn.Module):
    def __init__(self, bev_channels=256, embed_dim=256, traj_len=64, num_heads=4):
        super().__init__()
        self.traj_len = traj_len
        self.embed_dim = embed_dim
        
        # Time Step Embedding Generator
        self.time_mlp = nn.Sequential(
            TimestepEmbedding(embed_dim),
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, embed_dim)
        )
        
        # Action Input Projection (Maps continuous [acceleration, curvature] pairs)
        self.action_proj = nn.Linear(2, embed_dim)
        
        # BEV Image Feature Processing 
        # Compresses the spatial map or downsamples if map resolution is massive
        self.bev_downsample = nn.Sequential(
            nn.Conv2d(bev_channels, embed_dim, kernel_size=3, stride=1, padding=1),
            nn.GroupNorm(32, embed_dim),
            nn.SiLU()
        )
        
        # Attention Layer: Action queries look over the entire spatial BEV image map
        self.cross_attn = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
        
        # Layer Normalization Blocks
        self.norm_bev = nn.LayerNorm(embed_dim)
        self.norm_action = nn.LayerNorm(embed_dim)
        
        # Adaptive Layer Normalization (AdaLN) for condition scaling via time injection
        self.adaln_gamma = nn.Linear(embed_dim, embed_dim)
        self.adaln_beta = nn.Linear(embed_dim, embed_dim)
        
        # Prediction Output Head: Regresses the velocity field vectors (da/dt, d_kappa/dt)
        self.velocity_head = nn.Sequential(
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, 2)
        )

    def forward(self, u_t, t, bev_features):
        """
        u_t: Noisy control actions at flow time step t. Shape: (B, Traj_Len, 2) -> [accel, curvature]
        t: Continuous flow-matching time step. Shape: (B,) scalar tensor between [0.0, 1.0]
        bev_features: Image feature tensor from your BEV module backbone. Shape: (B, C, H, W)
        """
        B, _, H, W = bev_features.shape
        
        # 1. Process and encode continuous flow time
        t_embed = self.time_mlp(t) # Shape: (B, embed_dim)
        
        # 2. Flatten 2D BEV features into an attention-ready sequence
        processed_bev = self.bev_downsample(bev_features) # Shape: (B, embed_dim, H, W)
        bev_flat = processed_bev.permute(0, 2, 3, 1).reshape(B, H * W, self.embed_dim)
        bev_context = self.norm_bev(bev_flat) # Shape: (B, H*W, embed_dim)
        
        # 3. Project continuous action trajectories to Query space
        action_query = self.norm_action(self.action_proj(u_t)) # Shape: (B, Traj_Len, embed_dim)
        
        # 4. Perform cross-attention feature fusion
        # Queries (Actions) search through Keys & Values (BEV Context)
        attn_out, _ = self.cross_attn(query=action_query, key=bev_context, value=bev_context)
        
        # 5. Inject time embeddings using AdaLN modulation
        gamma = self.adaln_gamma(t_embed).unsqueeze(1) # Shape: (B, 1, embed_dim)
        beta = self.adaln_beta(t_embed).unsqueeze(1)   # Shape: (B, 1, embed_dim)
        fused_features = attn_out * (1 + gamma) + beta
        
        # 6. Regress Vector Field Velocity
        v_theta = self.velocity_head(fused_features) # Shape: (B, Traj_Len, 2)
        return v_theta

### PR DESCRIPTION
**Pull request by gcordova10**
## Description
This PR introduces an optional, swappable Bézier trajectory planner head into the `AutoE2E` architecture, aligning with the WG's requirements for trajectory smoothing and jerk reduction.

### Key Changes
- Implemented `BezierPlanner` under `Model/model_components/trajectory_planning/` using a precomputed Bernstein basis (zero `scipy` dependency).
- Integrated `PLANNER_REGISTRY` to support modular, swappable planners.
- Maintained full backward compatibility with the existing `AutoE2E` contract (`trajectory, ego_hidden, future_visual_features`).
- Follows unicycle dynamics (acceleration, curvature) as per the project's contract.
- Verified with 15+ unit/integration tests ensuring gradient flow and output integrity.

### Testing
- Verified numerical stability of Bernstein expansion against analytical reference.
- CI passed successfully (lint + unit tests).

Closes #45 (Relates to #51)

WG Alignment Note: This implementation aligns with the "fast/reactive" planning layer discussed in the WG meeting. It serves as one of the metric trajectory planner heads identified as a
  requirement for the current reactive block, ensuring compatibility with the broader architectural goal of separating reactive trajectory planning from future "slow-thinking" reasoning modules.

### CODE REVIEW

**Review by m-zain-khawaja — CHANGES_REQUESTED** [review]
Hi @gcordova10 - this looks good to me. I noticed that you added a compute_planner_loss function. We already have a losses folder with a [trajectory loss](https://github.com/autowarefoundation/auto_e2e/tree/main/Model/model_components/losses) - could you please therefore remove the planner loss from your bezier_planner.py and if you need to add further losses, could you please add them under the losses folder.

Also, have you tried a forward pass test with your planner, do the layers work correctly?

**Comment by gcordova10** [comment]
Hi @zainkhawaja! 
I have force-pushed to update this branch. The Bezier Planner is now fully refactored to comply with the new `BasePlanner` interface (separating `forward` for inference and `compute_planner_loss` for training), and I've integrated it cleanly via the Planner Registry. 
*Note on the CI:* The `make typecheck` (mypy) step might fail on this branch due to the pre-existing typing issue in `map.py` I kept my branch strictly focused on the Bezier planner and didn't include a fix here.
The PR is ready for review!

**Review by m-zain-khawaja — APPROVED** [review]
Approved, thanks @gcordova10

**Comment by gcordova10** [comment]
@RyotaYamada @m-zain-khawaja — as suggested in #56, I put together a small **planner benchmark harness** that runs `gru`, `flow_matching` and `bezier` through the `PLANNER_REGISTRY` under identical conditions (same `embed_dim`/horizon/inputs/device).

**What it measures now (no checkpoint/dataset/sim needed):** inference latency, parameter count, and architectural smoothness on the `(accel, curvature)` output — `jerk = Var(Δaccel)`, `curvature = Var(Δcurvature)`.

Results (GPU, batch=1, **random weights** — see caveat):

| planner | params | latency p50 (ms) | jerk Var(Δaccel) | curv Var(Δcurv) |
|---|---|---|---|---|
| gru (autoregressive) | 829K | 43.3 | 1.4e-04 | 3.8e-04 |
| flow_matching (ODE) | 988K | 11.3 | 4.2e+00 | 2.0e+00 |
| **bezier (one-shot)** | **495K** | **0.39** | **5.0e-06** | **1.2e-06** |

So Bézier is ~28× faster than flow-matching and ~110× faster than GRU, the lightest, and the smoothest by construction (Bernstein basis) — relevant for the real-time R-Car target.

**Honest caveat:** these are untrained weights, so the smoothness/latency numbers are *architectural*; **ADE/FDE, off-road rate, collision and closed-loop stability are NOT comparable yet** — they need a trained checkpoint + dataset (KITScenes/L2D) + a simulator (NAVSIM/Bench2Drive). The harness has hooks for those.

@m-zain-khawaja — since you're leading this comparison, happy to fold the harness into this PR or move it to a separate one, whatever you prefer. I can also help wire the trained ADE/FDE + closed-loop metrics once the training loop (#78/#67) is in place.

**Comment by zainkhawaja** [comment]
Stop Sent from my iPhoneOn Jun 22, 2026, at 3:48 PM, Gabriela Cordova ***@***.***> wrote:﻿gcordova10 left a comment (autowarefoundation/auto_e2e#69)
@ryotayamada @m-zain-khawaja — as suggested in #56, I put together a small planner benchmark harness that runs gru, flow_matching and bezier through the PLANNER_REGISTRY under identical conditions (same embed_dim/horizon/inputs/device).
What it measures now (no checkpoint/dataset/sim needed): inference latency, parameter count, and architectural smoothness on the (accel, curvature) output — jerk = Var(Δaccel), curvature = Var(Δcurvature).
Results (GPU, batch=1, random weights — see caveat):



planner
params
latency p50 (ms)
jerk Var(Δaccel)
curv Var(Δcurv)




gru (autoregressive)
829K
43.3
1.4e-04
3.8e-04


flow_matching (ODE)
988K
11.3
4.2e+00
2.0e+00


bezier (one-shot)
495K
0.39
5.0e-06
1.2e-06



So Bézier is ~28× faster than flow-matching and ~110× faster than GRU, the lightest, and the smoothest by construction (Bernstein basis) — relevant for the real-time R-Car target.
Honest caveat: these are untrained weights, so the smoothness/latency numbers are architectural; ADE/FDE, off-road rate, collision and closed-loop stability are NOT comparable yet — they need a trained checkpoint + dataset (KITScenes/L2D) + a simulator (NAVSIM/Bench2Drive). The harness has hooks for those.
@m-zain-khawaja — since you're leading this comparison, happy to fold the harness into this PR or move it to a separate one, whatever you prefer. I can also help wire the trained ADE/FDE + closed-loop metrics once the training loop (#78/#67) is in place.

**Comment by m-zain-khawaja** [comment]
Thank you @gcordova10 - these results are encouraging. Could you please add this in a separate PR 👍
